towhee/dataframe/dataframe_v2.py

```diff
@@ -33,18 +33,41 @@ class DataFrame:
 
     def __init__(
         self,
+        columns: List[Tuple[str, Any]],
         name: str = None,
         data=None,
-        columns=None,
     ):
         self._name = name
-        self._len = 0
         self._sealed = False
-        self._lock = threading.Lock()
+
+        self._iterator_lock = threading.Lock()
+        self._it_id = 0
+        self._iterators = []
+        self._iterator_offsets = []
+        self._min_offset = 0
+
+        self._block_lock = threading.Lock()
+        self._blocked = {}
```

Comment:
Maybe we could put the `condition` into iterators, that means iterator has a `condition` and a `notify` function, if iterator found there is no data, it calls `condition` wait. In dataframe's `put`, we can call iter's `notify` to wake up the iter.

Reply:
Maybe the `Event` is enough.

Reply:
```python 
elif code == 'Index_OOB_Unsealed':
            if self._block:
                cv = df.notify_block(self._offset, self._batch_size)
                with cv:
                    cv.wait()
                return self.__next__()

            return None

```

How about this? If the iterator needs to block it registers itself that it needs to block. In the df there is a dict that holds keeps track of needed indexes and a corresponding condition. 

```python 
def notify_block(self, offset, count):
        index = offset + count
        if not self._blocked.get(index):
            self._blocked[index] = threading.Condition()
        return self._blocked[index]
```

And then during a put it checks if there exists a cv waiting on that value. 

```python
self._len += 1

            cv = self._blocked.pop(self._len, None)

            if cv is not None:
                cv.notfiy_all()
```